### DISCUSSION TITLE
meaning of token properties in zowe.config.json

### DISCUSSION THREAD

**Discussion by markklerks** [zowe/zowe-cli#1485] [General]
For using the API ML and tokens, there are a few elements available in zowe.config.json.
At least there are:
- basePath (under properties)
- tokenType (under properties)
- tokenValue (under secure)
How do these work (together) in executing CLI commands with such a profile?
I can't find any detailed documentation on these and want to make sure to populate them correctly.

An next question: when I fill my config with all these elements and also my Username and password, will the mechanics be like it uses the username password once to get a token and from there the token will be used?

**Comment by t1m0thyj** [comment]
This documentation may be helpful: https://docs.zowe.org/stable/user-guide/cli-using-integrating-apiml

To summarize the purpose of each property:
* `basePath` is usually defined on a service profile to specify the base path for API ML to navigate to it. This path can be found in the API ML catalog (for example, a common base path for z/OSMF is `/ibmzosmf/api/v1`).
* `tokenType` is usually defined on a base profile since the API ML allows the same token to be used for all your API ML compliant services. The value for an API ML token is `apimlAuthenticationToken`.
* `tokenValue` is defined alongside token type and is often stored as a secure value like a password rather than in plain text. This is an API ML Single Sign-On (SSO) authentication token obtained by logging in to the API ML with your user and password.

In most cases you shouldn't have to manually add user, password, tokenType, or tokenValue to your config.
The `zowe auth login apiml` command will prompt you for your username and password, retrieve a token, and store it securely in your config for future commands to use.
You can also use the `zowe config auto-init` command to automatically generate an entire config file from scratch including the base paths, but many API ML services do not integrate with this command yet.
